Add cert-manager documentation (#1317)
minio/docs: source/operations/cert-manager.rst

.. _minio-certmanager:

============
cert-manager
============

.. default-domain:: minio

.. contents:: Table of Contents
   :local:
   :depth: 1

TLS certificate management with cert-manager
--------------------------------------------

> This guide shows you how to install `cert-manager <https://cert-manager.io>`__ and set up the global ``Cluster Issuer``.

Reviewer: ```suggestion
This guide shows you how to install  `cert-manager <https://cert-manager.io>`__ and set up a simple global ``Cluster Issuer``.

Defer to the cert-manager documentation and your own organizations certificate requirements for more advanced configurations.
```

Author: Couple of tweaks, but otherwise taking.

cert-manager manages certificates within Kubernetes clusters.

> The MinIO Operator supports using cert-manager for certificates as an alternative to the MinIO Operator managing certificates for itself and its tenants.

Reviewer: ```suggestion
The MinIO Operator supports using cert-manager for managing and provisioning certificates as an alternative to the MinIO Operator managing certificates for itself and its tenants.
```

cert-manager obtains valid certificates from an ``Issuer`` or ``ClusterIssuer`` and can automatically renew certificates prior to expiration.

A ``ClusterIssuer`` issues certificates for multiple namespaces.
An ``Issuer`` only mints certificates for its own namespace.

The following graphic depicts how various namespaces make use of either an ``Issuer`` or ``ClusterIssuer`` type.

> - cert-manager is installed in the ``cert-manager`` namespace, which does not have either an ``issuer`` or a ``ClusterIssuer``.

Reviewer 2: I'm confused by the `cert-manager` namespace in the diagram, is `minio-operator:issuer` not an `issuer`?

Author: cert-manager is the ClusterIssuer.
It can issue certificates to MinIO Operator, tenants, and any other apps that might also run in the cluster, whether related or unrelated to MinIO.

Author: We can address that with other comments on the graphic.
It should not have either an Issuer or ClusterIssuer in that namespace. But it operates separately from the other namespaces.

Author: See suggestion in other thread here
https://github.com/minio/docs/pull/1317#discussion_r1765237318

> - cert-manager is installed in the ``cert-manager`` namespace, which does not have either an ``issuer`` or a ``ClusterIssuer``.

Reviewer: Do we still need this if we're dropping a reference to it in the diagram?

Maybe:

```
The following diagram provides a high level view of the relationship between Cluster Issuers and Issuers in a cert-manager context:

Author: Holding for the new diagram. Will leave unresolved until then.

- The ``default`` namespace receives the global ``Cluster Issuer``.
- Each tenant's namespace receives a local ``Issuer``.
- The ``minio-operator`` namespace receives a local ``Issuer``. 

.. image:: /images/k8s/cert-manager-cluster.svg
   :width: 600px

>    :alt: A Kubernetes cluster with five namespaces, shown as a box for each namespace in the cluster. The minio-operator namespace contains a "minio-operator: issuer" local issuer. The default namespace contains a "root: ClusterIssuer" cluster issuer. The cert-manager namespace contains a "minio-operator: issuer" local issuer. The remaining two namespaces are individual tenants, "tenant-1" and "tenant-2", each with its own local issuer.

Reviewer: Need to update the alt text

Author: Waiting for the new diagram first. Then will update.

Author: Alt text updated for the new diagram.

>    :alt: A graphic depiction of a kubernetes cluster with five separate namespaces represented by different boxes. One is labeled minio-operator with a text box that says "minio-operator: issuer". A second is labeled tenant-1 with text box that says "tenant-1: issuer". A third is labeled default with a text box that says "root:ClusterIssuer". A fourth is labeled tenant-2 with a text box that says "tenant-2: issuer". The fifth is labeled cert-manager with a text box that says "minio-operator: issuer".

Reviewer 2: An idea for the alt-text
```suggestion
   :alt: A Kubernetes cluster with five namespaces, shown as a box for each namespace in the cluster. The minio-operator namespace contains a "minio-operator: issuer" local issuer. The default namespace contains a "root: ClusterIssuer" cluster issuer. The cert-manager namespace contains a "minio-operator: issuer" local issuer. The remaining two namespaces are individual tenants, "tenant-1" and "tenant-2", each with its own local issuer.
```
(I'm not sure how screen readers handle quotes and terminal punctuation, so I followed your style.)

Author: Alt text is not my forte, so I'm open to whatever suggestions. I just took what you suggested.

Reviewer 2: Warning, opinionated digression ahead:

My take on alt-text is "What are the important ideas the author intended people to get from this image?" And everybody has their own opinion on how it works for them and their technology. For me, reading a description of every detail (on my tiny communications device) makes it harder to follow the purpose of the image.

Lately I've been seeing more literal image descriptions and I wonder if they are auto generated. For example, computers can tell you the image has overlapping circles containing text, but perhaps not that it's a Venn diagram illustrating a joke.

   :align: center

.. note::

   This guide uses a self-signed ``Cluster Issuer``. 
   You can also use `other Issuers supported by cert-manager <https://cert-manager.io/docs/configuration/issuers/>`__.

>    The main difference is that you must provide the ``Issuer`` CA certificate to MinIO, instead of the CA's mentioned in this guide.

Reviewer: Move this up to where we talk about what the guide is doing?

Also struggling with the sentence on 43. Is it because with this guide, we would use the ClusterIssuer for MinIO instead of an Issuer?

Author: It's a matter of who the issuer is.
So, if you're using a different issuer as their docs mention, then use those certificates instead of the step we have for issuing and trusting certificates here. Basically, if you use another issuer, you're on your own.

Author: Moved the admonition and added a `that` to what was line 43 to hopefully make it clearer.

.. _minio-certmanager:

============
cert-manager
============

.. default-domain:: minio

.. contents:: Table of Contents
   :local:
   :depth: 1

TLS certificate management with cert-manager
--------------------------------------------

This guide shows you how to install cert-manager for TLS certificate management.
The guide assumes a new or fresh MinIO Operator installation.

.. note::

   This guide uses a self-signed ``Cluster Issuer``. 
   You can also use `other Issuers supported by cert-manager <https://cert-manager.io/docs/configuration/issuers/>`__.

   The main difference is that you must provide that ``Issuer`` CA certificate to MinIO, instead of the CA's mentioned in this guide.

> Defer to the `cert-manager documentation <https://cert-manager.io>`__ and your own organization's certificate requirements for more advanced configurations.

Reviewer 3: ```suggestion
Refer to the `cert-manager documentation <https://cert-manager.io>`__ and your own organization's certificate requirements for more advanced configurations.
```
?

cert-manager manages certificates within Kubernetes clusters.
The MinIO Operator supports using cert-manager for managing and provisioning certificates as an alternative to the MinIO Operator managing certificates for itself and its tenants.

cert-manager obtains valid certificates from an ``Issuer`` or ``ClusterIssuer`` and can automatically renew certificates prior to expiration.

A ``ClusterIssuer`` issues certificates for multiple namespaces.
An ``Issuer`` only mints certificates for its own namespace.

The following graphic depicts how cert-manager provides certificates in namespaces across a Kubernetes cluster.

- A ``ClusterIssuer`` exists at the root level of the Kubernetes cluster, typically the ``default`` namespace, to provide certificates to all other namespaces.
- The ``minio-operator`` namespace receives its own, local ``Issuer``. 
- Each tenant's namespace receives its own, local ``Issuer``.
- The certificates issued by each tenant namespace must be made known to and trusted by the MinIO Operator.

.. image:: /images/k8s/cert-manager-graph.png
   :width: 600px

>    :alt: A graph of the namespaces in a Kubernetes cluster. At the root level is a ClusterIssuer that issues certificates to each of the other namespaces. Three other namespaces exist under root. The Operator namespace has its own Issuer, as do namespaces for Tenant-1 and Tenant-2. A dotted line going from each Tenant namespace back ot the Operator has text that the Operator must trust the Tenant certificates.

Reviewer 3: ```suggestion
   :alt: A graph of the namespaces in a Kubernetes cluster. At the root level is a ClusterIssuer that issues certificates to each of the other namespaces. Three other namespaces exist under root. The Operator namespace has its own Issuer, as do namespaces for Tenant-1 and Tenant-2. A dotted line going from each Tenant namespace back to the Operator has text that the Operator must trust the Tenant certificates.
```
Typo fix. Question, though: the description of the graphic looks really good, and it's my understanding that in this kind of situation, effectively duplicate alt text not only isn't required, it's actually not so good for accessibility. Maybe something for larger discussion rather than addressing here, but figured I'd call it out.

Author: Hmm. That's a good point. I've simplified the alt text.

.. _minio-certmanager:

============
cert-manager
============

.. default-domain:: minio

.. contents:: Table of Contents
   :local:
   :depth: 1

> TLS certificate management with cert-manager

Reviewer: it looks weird as heck but should we `cert-manager` the header?

Reviewer 2: Their own docs use "cert-manager" in body text. Lowercase even at the beginning of a sentence.

--------------------------------------------

`cert-manager <https://cert-manager.io/>`__ works as a controller for managing certificates within Kubernetes clusters.

> cert-manager obtains valid certificates from an ``Issuer`` or ``ClusterIssuer`` and can renew certificates prior to expiration.

Reviewer: Hm - sphinx isn't flexible in monospacing the HTML link, but the actual reference should probably be `cert-manager`?

Author: I don't think we need to monospace the product name.

> cert-manager obtains valid certificates from an ``Issuer`` or ``ClusterIssuer`` and can renew certificates prior to expiration.

Reviewer: ```suggestion
cert-manager obtains valid certificates from an ``Issuer`` or ``ClusterIssuer`` and can automatically renew certificates prior to expiration.
```

Author: Taken!

The MinIO Operator supports using cert-manager for certificates as an alternative to the MinIO Operator managing certificates for itself and its tenants.

On a Kubernetes cluster, cert-manager requires a global level ``ClusterIssuer`` to generate intermediary ``Issuers`` and certificates.
At the namespace level, cert-manager issues certificates derived from an ``Issuer``.
A ``ClusterIssuer`` can issue certificates for multiple namespaces.
An ``Issuer`` only issues certificates for its own namespace.

To learn more about cert-manager Issuers, refer to the `Issuer documentation <https://cert-manager.io/docs/concepts/issuer/>`__.

> Below is a logical view of a Kubernetes cluster with four namespaces:

Reviewer: I'm not sure we need this for the purpose of the tutorial. We're also bringing in 'STS' in the diagram and i don't think that wins us any points here.

Author: Cut.

- ``minio-operator``
- ``tenant-1``
- ``tenant-2``

> - ``other-namespace``

Reviewer: We never reference this again, i'm not sure it does much for us. We can limit scope to just how cert manager works w/ minio resources.

Author: I removed this entirely in an IA rework of the page into multiple pages.

.. image:: /images/k8s/cert-manager-namespaces.svg
   :width: 600px

>    :alt: A graphic depiction of a kubernetes cluster with four namespaces. Each namespace in its own white box. One is labeled minio-operator with contents two sts content items. A second is labeled tenant-1 with a MinIO logo and four drives of a pool. A third is labeled tenant-2 with similar contents to the tenant-1 box. The fourth is labeled other namespace with contents of other pods.

Reviewer 2: I've got some ideas about diagram alt-text, but that can wait until we figure out more on what should be in these. Leaving this comment as a reminder for myself to follow up. @feorlen

   :align: center

> This guide shows you how to set up a different Certificate Authority (CA) in each namespace.

Reviewer: I would jump straight to this.

Author: Moved to the top of the page.

Author: But keeping a chunk of the rest of the content for context. Because you need to install cert-manager, then install a certificate for cert-manager, the operator, and each tenant. Trying to explain all of that to set explectations.

Each namespace's CA references the global ``Cluster Issuer``.

The following graphic depicts how various namespaces make use of either an ``Issuer`` or ``ClusterIssuer`` type.

- cert-manager is installed in the ``cert-manager`` namespace, which does not have either an ``issuer`` or a ``ClusterIssuer``.
- The ``default`` namespace receives the global ``Cluster Issuer``.
- Each tenant's namespace receives a local ``Issuer``.
- The ``minio-operator`` namespace receives a local ``Issuer``. 
  More about services that require TLS certificates in the ``minio-operator`` namespace are covered :ref:`below <minio-operator-services-with-cert-manager>`.

> .. image:: /images/k8s/cert-manager-cluster.svg

Reviewer: Suggest modifying this to be a root-leaf type diagram where certmanager is the 'root' and then each 'branch' leads to the leaf namespaces.

Author: I guess the question is do we want to show how cert-manager works from the view point of the cluster, or how the certificates work within a cert-manager world.

Author: Something more like this?

![certmanagerissuers drawio](https://github.com/user-attachments/assets/a20ffbea-c480-47dc-a7d5-179e6c0bbe48)

Author: I think for the purposes of our docs, the focus should be on the ClusterIssuer > Issuer relationship. The original diagram was focused more on namespaces.

Reviewer: Agreed - less is more, and I think focusing on how cert-manager works with MinIO + Tenants is the critical bit.

   :width: 600px
   :alt: A graphic depiction of a kubernetes cluster with five separate namespaces represented by different boxes. One is labeled minio-operator with a text box that says "minio-operator: issuer". A second is labeled tenant-1 with text box that says "tenant-1: issuer". A third is labeled default with a text box that says "root:ClusterIssuer". A fourth is labeled tenant-2 with a text box that says "tenant-2: issuer". The fifth is labeled cert-manager with a text box that says "minio-operator: issuer".
   :align: center

.. note::

   This guide uses a self-signed ``Cluster Issuer``. 
   You can also use `other Issuers supported by cert-manager <https://cert-manager.io/docs/configuration/issuers/>`__.
   The main difference is that you must provide the ``Issuer`` CA certificate to MinIO, instead of the CA's mentioned in this guide.

Prerequisites
-------------

- A `supported version of Kubernetes <https://kubernetes.io/releases/>`__. 
 

>   While cert-manager supports `earlier K8s versions <https://cert-manager.io/docs/installation/supported-releases/>`__, the MinIO Operator requires a currently supported version.

Reviewer: I'd drop this. We state what the supported version is, that's all we need.

Author: Dropped.

- `kustomize <https://kustomize.io/>`__ installed
- ``kubectl`` access to your ``k8s`` cluster

Getting Started
---------------

> 

Reviewer: This feels odd - too many H1s, and this isn't a 'Getting Started' it's basically Step 1: Set up Cert Manager.

Suggest:

H2: Set Up Cert Manager

H3: Install certmanager

H3: Configure self-signed Cluster issuer

H2: Configure MinIO Operator to use cert-manager

H3: stuff

H3: more stuff

etc.

Author: Wound up splitting the page up into multiple pages. I think it works better.

Setup cert-manager
~~~~~~~~~~~~~~~~~~

Install cert-manager. 
`Release 1.12.X LTS <https://cert-manager.io/docs/releases/release-notes/release-notes-1.12/>`__ is preferred, but you may install latest.

> The following command installs version 1.12.13 using ``kubectl``.

Reviewer: Worth doing a substitution here in case the version bumps later?

Author: Perhaps? Not sure how critical that is to stay up. Mostly we want this lineage. Which particular release within 1.12.x is less important from our perspective.

      :class: copyable

      kubectl apply -f selfsigned-root-clusterissuer.yaml

.. _minio-operator-services-with-cert-manager:

MinIO Operator services with cert-manager
-----------------------------------------

MinIO Operator manages the TLS certificate issuing for the services hosted in the ``minio-operator`` namespace. 
That is the :ref:`Secure Token Service (sts) <minio-sts-operator>`.

This section describes how to generate the ``sts`` TLS certificate with cert-manager.

- These certificates **must** be issued *before* installing Operator.
- The cluster's self-signed root ``ClusterIssuer`` certificate must already exist, as described above.

Secure Token Service (STS)
~~~~~~~~~~~~~~~~~~~~~~~~~~

MinIO STS is a service included with MinIO Operator that provides Native IAM Authentication for Kubernetes. 
This service allows you to control access to your MinIO tenant from your kubernetes applications without having to explicitly create credentials for each application. 
For more information, see the :ref:`Secure Token Service documentation <minio-sts-operator>`.

Think of the STS Service as a webserver presented with a TLS certificate for ``https`` traffic.

> This guide covers how to **disable** the automatic generation of the certificate in MinIO Operator and issue the certificate using cert-manager instead.

Reviewer: So I'm not super familiar with this, but is it _absolutely required_ to talk about the STS certificate here? I feel like it will confuse users, when all they really need to know is "If you set up cert-manager for MinIO Operator, you can use those certificates for authenticating to the Operator" or something.

Author: I cut out most of this.

   Create a file named ``sts-tls-certificate.yaml`` with the following contents:

   .. code-block:: yaml
      :class: copyable
      :emphasize-lines: 7,12

      # sts-tls-certificate.yaml
      apiVersion: cert-manager.io/v1
      kind: Certificate
      metadata:
        name: sts-certmanager-cert
        namespace: minio-operator
      spec:
        dnsNames:
          - sts
          - sts.minio-operator.svc
          - sts.minio-operator.svc.cluster.local
        secretName: sts-tls
        issuerRef:
          name: minio-operator-ca-issuer

   .. important::
   
      The ``spec.secretName`` is not optional! 
   
      The secret name **must be** ``sts-tls``.
      Confirm this by setting ``spec.secretName: sts-tls`` as highlighted above.

2. Apply the resource to the cluster:

   .. code-block:: shell
      :class: copyable

      kubectl apply -f sts-tls-certificate.yaml

This creates a secret called ``sts-tls`` in the ``minio-operator`` namespace.

.. warning::
  
   Failing to provide the ``sts-tls`` secret containing the TLS certificate or providing an invalid key-value pair in the secret will prevent the STS service from starting.

> Install Operator with Auto TLS disabled for STS

Reviewer: Same thought - this is _cool_, but I cannot shake the feeling that what users want to do is set up Tenants with cert-manager.

Configuring Operator to use cert-manager for STS authentication for <reasons> is more advanced. I think it should be its own sub-tutorial, unless I'm missing something.

Author: So I tried to minimize the STS language. I *think* STS is the underlying thing going on in the background, though. You have to disable Auto TLS to get cert-manager to work. So that part has to stay.
